### PR TITLE
Add `set_encoding`/`unset_encoding` operations to `LinalgExt`

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
The `iree_linalg_ext.pack` and `iree_linalg_ext.unpack` operations
represent the change in data layout by explicitly changing the rank of
the tensors. This might be useful for code-generation, but higher up
the stack results of these operations will then need special tiled
versions of operations like `linalg.matmul` and `linalg.conv_*`.
Instead, the ops added here represent the layout change as a tensor
encoding without changing the logical shape of the tensor. These pack
and unpack operations then can be lowered to pack and unpack
operations that explicitly introduce the higher dimensional tensors
within a dispatch.

The encoding attribute used can be richer, but for now they are just
an enum with semantic meaning that can be used as required. Over time
these encoding attributes can be modified to be more explicit in
describing the layout change to perform.

### CODE REVIEW

**Review by benvanik — COMMENTED** [review]
neat!
please name flow.tensor.pack/unpack to match the other tensor ops we have

**Comment by MaheshRavishankar** [comment]
@benvanik as PR #10481 makes progress, I was trying to see what we need at Flow level so that outside of dispatches there is enough information for connecting this e2e (through Stream and HAL dialects). As I understand it, we dont need any runtime modifications.

There are many different solutions, and I dont want to over engineer things. So starting with the minimal possible footprint at Flow level that allows lowering to Stream and HAL such that within dispatches we can lower these to the `pack` and `unpack` operations being added in the above PR. So starting with a simple annotation. If this is enough we can build up a whole family of annotations, and have a better attribute to describe the packing we want. Sparse tensor for example has a rich encoding attribute that we can mimic (https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/SparseTensor/roundtrip_encoding.mlir). 

If this is not enough, we can go all in and try `TypeInterface`s for the tiled type, etc. but not sure we want to go down that path without more data. Let me know.

**Comment by benvanik** [comment]
(typing this up here but happy to chat more/whiteboard/etc - just think this is a complex enough topic that I needed to try typing this a half dozen times to feel it out first and develop feelings about things :)

I still don't think this will work for us - we need a unique type when exiting the flow level even if that doesn't make it into the linalg logic inside dispatch regions. So maybe this is one pair of ops used after dispatch region formation inside of dispatches but I think we'll need the others and I don't see how to directly lower this to those today - though maybe it unblocks codegen enough to land anyway? I'd probably want it renamed to experimental or something if it did land as it sketches me out a bit :)

Thinking about it: for whole-program stuff at the flow level I believe reusing the naked tensor type (no encoding) here makes things too brittle (we can't perform a lot of operations on these tensor values but won't know that based on just the ops - function args, stored globals, values passed in from users etc all look the same), discards too much information (we don't know when something is used in a certain form after we create dispatch regions, and thus can't assign devices and insert transfers where required), and forces a level of dynamic behavior we can't reasonably handle lower down (all those extra ? have real cost on allocation, extra runtime tracking, and the potential value set information we can provide to dispatches). I don't see how any flow pass (or subsequent stream pass) could run on the tensors produced/consumed by these ops which makes me worry that it'd be a bug farm for you when they don't get put in dispatch regions.

That doesn't mean that within formed dispatch regions we couldn't do this but only that at the level of whole-program stuff (IPO, elision, etc) before dispatch region formation we need something different and after dispatch region formation on the outside we need something different. Maybe these two before/after cases could be the same and we end up with 2 op pairs total (one for outside and one for inside) or maybe we end up with 3 (pre/inside/post). If we had the pre-formation op and then during formation lowered that to the post-formation op on the outside and dispatch-local op (like the ones here) that'd probably be fine and not too complex: it's a nice unidirectional lowering operation as we partition things.

I think the core requirements are that before dispatch region formation we can't discard the logical shape information (original rank/dims) and can't introduce the unknowable device-specific information (innermost 2D dims), while after dispatch region formation we need to know whether each tensor SSA value is special and what can be done to it and use that to assign devices/introduce transfers/etc. The unspecialized `tensor` type doesn't let us do those things, but is probably fine for what's inside dispatch regions as it interops with linalg and other dialects.

I _think_ the easiest/most generalizable solution is to use `tensor` with a custom encoding: pack goes from dense linear tensors to packed ones and unpack does the opposite, changing only the encoding and not the dimensionality. That's because at the level of pre-flow IPO we care only about the dimensions we can reason about (the original ones) and after flow we must know how to reason about the size and compatibility of any tensor value (using its original logical size and its encoding). Another option would be to have our own type but I'm not sure how that'd look different than `tensor` with an encoding attribute - and we need to support tensors with alternative encodings (sparsity/etc) anyway so a custom type that lowers into stream would just be more work. I stand by that all this packing/padding/etc is identical in form to sparsity and that thinking about it like that makes things easier so I'd hate to see a divergent path.

If we had only one op it'd be something like the ones proposed here pre-flow and we'd need to ensure none exist after dispatch region formation. I don't quite see how to ensure that after we perform canonicalization/IPO/consteval/etc unless we very carefully manage the pipelines. It also wouldn't work for heterogenous or anything else we need to do in the stream dialect as it'd be sequestered in dispatch regions and introducing (or even detecting the need for) repacking is not possible. Ops are cheap and having one that survived after dispatch region formation (even if normally we try to fuse) is going to be the most robust and work best with the system where we need to implement tied/shape-aware ops anyway.

If we had a second op to cover the post dispatch region formation time I suspect it'd look like:

```mlir
  %0 = flow.tensor.pack %arg0 : tensor<?x?xf32>{%d0, %d1} -> tensor<?x?xf32, #gemm_out>{%d0, %d1}
  %1 = flow.tensor.unpack %0 : tensor<?x?xf32, #gemm_out>{%d0, %d1} -> tensor<?x?xf32>{%d0, %d1}
```

And then the simpler rank-changing op proposed that doesn't capture dims would be exclusively used in dispatch regions on the result of a `flow.dispatch.tensor.load` or operand of a `flow.dispatch.tensor.store` (or anything inbetween). Dispatch regions could then take/return the tensor with custom encoding and if any of those ops survived they'd be lowered to builtins in the stream dialect (dispatches that would then just contain pack/unpack ops). It's the job of the stream dialect partitioning code to then ensure packed tensors never leave span differing devices and insert packs/unpacks as needed in order to satisfy that. So after assignment it may end up like this in the stream dialect:

```mlir
  %dev_a = stream.async.dispatch affinity(#dev_a) ... -> tensor<?x?xf32, #gemm_out>{%d0, %d1} in %resource ...
  %linear = stream.tensor.unpack affinity(#dev_a) ... %dev_a : tensor<?x?xf32, #gemm_out>{%d0, %d1} in %resource ... -> tensor<?x?xf32>{%d0, %d1} in %resource ...
  %dev_b = stream.tensor.pack affinity(#dev_b) ... %linear : tensor<?x?xf32>{%d0, %d1} in %resource ... -> tensor<?x?xf32, #gemm_out>{%d0, %d1} in %resource ...
  stream.async.dispatch affinity(#dev_b) ... (%dev_b)
```
and then it's a simple optimization pass to fuse in the pack/unpack into the `flow.dispatch.tensor.load`/store ops. Whatever op exists on the inside (maybe the proposed pack/unpack ops here) would work.

Note that the above example is identical to sparsity from this perspective of needing to pack/unpack, transfer, etc with `#gemm_out` being replaced with `#csr` or whatnot - the required information is available (logical tensor size, the encoding, and the affinity) but the details are hidden (what's the exact 2D inner tile size on a particular device or exact bit layout/blocking of the sparse data) inside of dispatches.

So all of that is to say: I think this may work inside dispatch regions but wouldn't be usable outside before forming them or outside after forming them. I don't think the ones that are usable outside are too wild, and probably look just like a renamed `flow.tensor.clone` that can have differing encodings on args/results.

**Comment by MaheshRavishankar** [comment]
@benvanik I hope you are looking at the ops defined in this draft and not the ones defined in #10481 . That op is for tie-ing things for codegen. In this PR I am adding ops to the Flow level for "outside  of codegen" use case. Having said that, 

> ```mlir
>  %0 = flow.tensor.pack %arg0 : tensor<?x?xf32>{%d0, %d1} -> tensor<?x?xf32, #gemm_out>{%d0, %d1}
>   %1 = flow.tensor.unpack %0 : tensor<?x?xf32, #gemm_out>{%d0, %d1} -> tensor<?x?xf32>{%d0, %d1}
> ```

is pretty close. Instead of `#gemm_out` I just added a `"gemm_out"`, ie. a `StringAttr` instead of an explicitly defined encoding. We can have an encoding in due course. That matches well with the direction Sparse codegen is taking. I think I went with generating a `tensor<?x?x?x?xf32>` since I was thinking of how this would connect to `linalg.mmt4d`, but we dont need to do this "outside" of codegen. I suspect within the dispatch we can lower the `flow.tensor.pack` to `linalg_ext.pack` operation that is being defined in #10481 and also lower to `linalg.mmt4d`. (Maybe we dont even need to use `linalg.mmt4d` have to see). 

Summary : Instead of going from a tensor of rank `N` to a tensor of rank `N + tile_dims`. we use the `flow.tensor.pack` operation to just add a tensor encoding while staying in the same rank. That sounds like a good starting point if you think that can connect properly to Stream/HAL, etc.

**Comment by benvanik** [comment]
Sounds good! I'm happy to connect up the stream stuff or help with the flow ops!

**Comment by MaheshRavishankar** [comment]
> Listening! If I understand correctly, if you provided a lowering of this flow pack op to the linalg-ext pack op from #10481 , then I could start using this instead of the current expandshape+transpose in the matmul-to-mmt4d pattern.
> 

Yeah, I am trying to first get an e2e of just the `linalg_ext.pack` working (hopefully by EoD today).  This is only `pack` for now. There is an `unpack` coming... I was waiting for those to land to start having the path of doing data tiling for GEMMs truly working in IREE.

> I like that such a switch of matmul-to-mmt4d to flow.pack can be done immediately while the delicate question of how we populate those target-specific `?` values in the 4d shapes can be dealt with orthogonally to that.
> 
> I understand from #10481 that that doesn't do padding yet. The flow.pack definition in the present PR has a comment saying that it does pad. I hope that the plan is to define padding semantics in flow.pack and move the padding from matmul-to-mmt4d to flow-pack-to-linalgext-pack, right? Until linalgext-pack gets padding semantics itself?

The padding to `linalg_ext.pack` is being looked into by Hanhan. Expect it to be there EoW. I do need to add padding to `flow.pack`. Will do so when it is time to connect things e2e.

**Comment by MaheshRavishankar** [comment]
@benvanik updated based on discussion above. PTAL.

Had a discussion with @nicolasvasilache as well. He suggested moving this to `iree_linalg_ext` as well since it might be something worth pushing upstream.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp [hunk 1]
```diff
@@ -82,10 +83,20 @@ static Value getSlice(OpBuilder &b, Location loc, Value source,
...
-static bool isShapedTypeDimEqual(int64_t lhs, int64_t rhs) {
-  return lhs != ShapedType::kDynamicSize && rhs != ShapedType::kDynamicSize &&
-         lhs != rhs;
+/// Returns true if the dimensions of ShapedType are compatible
+static bool isShapedTypeDimCompatible(int64_t lhs, int64_t rhs) {
+  return lhs == ShapedType::kDynamicSize || rhs == ShapedType::kDynamicSize ||
+         lhs == rhs;
+}
+
+/// Returns true if the dimensions of ShapedType arent dynaim
```
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:92]: complete the comment

**Review by benvanik — APPROVED** [review]
Out of curiosity, why set/unset instead of a single op ala cast? (wondering if there's representational advantages/disadvantages to repacking via unset+set)

**Comment by MaheshRavishankar** [comment]
> Out of curiosity, why set/unset instead of a single op ala cast? (wondering if there's representational advantages/disadvantages to repacking via unset+set)

Yeah, that'd work too, but starting with something explicit, mostly cause logic of when to lower to pack and unpack operations becomes simpler. Not sure it justifies it fully. Ill keep it in mind if it is worth collapsing into a single op.

**Pull request merged**